### PR TITLE
Add implementation for Distribution API

### PR DESCRIPTION
**Pull request by qlzh727**
1. Implemented DeviceMesh and TensorLayout
2. Implemented list_devices.
3. Add backend functions for list_devices, and conversion functions to jax native Mesh and Sharding.
4. Update the TensorLayout to also include the DeviceMesh, which is needed when converting the actual backend layout.
5. Added unit tests.

Note that there is a dummy distribution_lib in the backend/__init__.py, the unit test will fail for other backends (due to import) if this dummy placeholder is not there.

### CODE REVIEW

File: keras_core/backend/jax/distribution_lib.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+def list_devices(device_type=None):
+    """Return all the available devices based on the device type.
+
+    Note that this should return the global devices in a distributed setting.
+
+    Args:
+        device_type: string of 'CPU', 'GPU' or 'TPU'. Default to GPU or TPU if
+            available when device_type is not provided. Otherwise will return
+            the CPU devices.
```
Comment (fchollet) [inline keras_core/backend/jax/distribution_lib.py:19]: QQ: is defaulting to CPU consistent with what happens with JAX and TF list_device?

File: keras_core/backend/jax/distribution_lib.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    Args:
+        device_mesh: DeviceMesh instance to convert.
+
+    Returns:
+        A `jax.sharding.Mesh` instance.
+    """
+    return jax.sharding.Mesh(device_mesh.devices, device_mesh.axis_names)
+
+
+def to_backend_layout(tensor_layout):
```
Comment (fchollet) [inline keras_core/backend/jax/distribution_lib.py:40]: Same here -- every backend can have its own `to_xxx_layout` (it will only be used inside the backend anyway)

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_core/backend/jax/distribution_lib.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import jax
+
+
+def list_devices(device_type=None):
+    """Return all the available devices based on the device type.
+
+    Note that this should return the global devices in a distributed setting.
+
+    Args:
+        device_type: string of 'CPU', 'GPU' or 'TPU'. Default to GPU or TPU if
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py]: Done.

File: keras_core/backend/jax/distribution_lib.py (hunk 1 above)
Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py]: Yes, the priority order of device type is "GPU" == "TPU" > "CPU", when the device type is not specified. 

This is same between JAX and tf.dtensor.

File: keras_core/backend/jax/distribution_lib.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            the CPU devices.
+
+    Return:
+        List of devices that are available for distribute computation.
+    """
+    device_type = device_type.lower() if device_type else "cpu"
+    return jax.devices(backend=device_type)
+
+
+def to_backend_mesh(device_mesh):
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py]: Done.

File: keras_core/backend/jax/distribution_lib.py (hunk 2 above)
Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py]: Done.

File: keras_core/backend/jax/distribution_lib.py (hunk 1 above)
Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py]: I see your question, I was incorrectly converting None to "CPU", and it has been changed back now.

Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py]: Btw, I don't think we have GPU/TPU test setup for OSS build. Depends on how hard to hookup the TPU/GPU with github actions, we might also need to consider setting the device test internally.

File: keras_core/backend/jax/distribution_lib.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    Args:
+        device_type: string of `"CPU"`, `"GPU"` or `"TPU"`. Default to `GPU` or
+            `TPU` if available when device_type is not provided. Otherwise
+            will return the `CPU` devices.
+
+    Return:
+        List of devices that are available for distribute computation.
+    """
+    device_type = device_type.lower() if device_type else None
```
Comment (fchollet) [inline keras_core/backend/jax/distribution_lib.py:24]: This means that JAX will be resolving its own defaults. This is fine, but we'll need to make sure we get the same default behavior in all backends, when we add TF etc.

File: keras_core/backend/jax/distribution_lib.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    Args:
+        device_mesh: DeviceMesh instance to convert.
+
+    Returns:
+        A `jax.sharding.Mesh` instance.
+    """
+    return jax.sharding.Mesh(device_mesh.devices, device_mesh.axis_names)
+
+
+def to_jax_layout(tensor_layout):
```
Comment (fchollet) [inline keras_core/backend/jax/distribution_lib.py:40]: Are you able to add `distribute_tensor(tensor, layout)` at this time?

File: keras_core/backend/jax/distribution_lib.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import jax
+
+
+def list_devices(device_type=None):
+    """Return all the available devices based on the device type.
+
+    Note that this should return the global devices in a distributed setting.
+
+    Args:
+        device_type: string of `"CPU"`, `"GPU"` or `"TPU"`. Default to `GPU` or
```
Comment (fchollet) [inline keras_core/backend/jax/distribution_lib.py:17]: We normalize to lowercase and we generally use lowercase for string constants everywhere, so let's them lowercase in the docstring

File: keras_core/distribution/distribution_lib.py [hunk 8]
```diff
@@ -19,14 +21,14 @@ def list_devices(device_type=None):
     Note that this should return the global devices in a distributed setting.
 
     Args:
-        device_type: string of 'CPU', 'GPU' or 'TPU'. Default to GPU or TPU if
-            available when device_type is not provided. Otherwise will return
-            the CPU devices.
+        device_type: string of `"CPU"`, `"GPU"` or `"TPU"`. Default to `GPU` or
```
Comment (fchollet) [inline keras_core/distribution/distribution_lib.py:24]: Here too

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras_core/backend/jax/distribution_lib.py (hunk 7 above)
Comment (qlzh727) [reply keras_core/backend/jax/distribution_lib.py:17]: i see. tf.dtensor was using upper case for device type, but jax was using lower case.

File: keras_core/distribution/distribution_lib.py (hunk 8 above)
Comment (qlzh727) [reply keras_core/distribution/distribution_lib.py:24]: Done.

**Review by fchollet — APPROVED** [review]
LGTM
